cmd/expiration-mailer/main_test.go

```diff
@@ -289,6 +290,152 @@ func TestFindExpiringCertificates(t *testing.T) {
+	ctx := setup(t, []time.Duration{time.Hour * 24, time.Hour * 24 * 4, time.Hour * 24 * 7})
+
+	log.Clear()
+
+	emailA, _ := core.ParseAcmeURL("mailto:one@mail.com")
+	var keyA jose.JsonWebKey
+	err := json.Unmarshal(jsonKeyA, &keyA)
+	test.AssertNotError(t, err, "Failed to unmarshal public JWK")
+	regA := core.Registration{
+		ID: 1,
+		Contact: []*core.AcmeURL{
+			emailA,
+		},
+		Key:       keyA,
+		InitialIP: net.ParseIP("2.3.2.3"),
+	}
+	regA, err = ctx.ssa.NewRegistration(regA)
+	if err != nil {
+		t.Fatalf("Couldn't store regA: %s", err)
+	}
+
+	testCerts := []*struct {
+		Serial       int
+		stringSerial string
+		FQDNHash     []byte
+		DNS          string
+		NotBefore    time.Time
+		NotAfter     time.Time
+		// this field is the test assertion
+		IsRenewed bool
+	}{
+		{
+			Serial:    1001,
+			FQDNHash:  []byte("hash of A"),
+			DNS:       "a.example.com",
+			NotBefore: ctx.fc.Now().Add((-1 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((89 * 24) * time.Hour),
+			IsRenewed: true,
+		},
+		{
+			Serial:    1002,
+			FQDNHash:  []byte("hash of A"),
+			DNS:       "a.example.com",
+			NotBefore: ctx.fc.Now().Add((0 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((90 * 24) * time.Hour),
+			IsRenewed: false,
+		},
+		{
+			Serial:    1003,
+			FQDNHash:  []byte("hash of B"),
+			DNS:       "b.example.net",
+			NotBefore: ctx.fc.Now().Add((0 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((90 * 24) * time.Hour),
+			IsRenewed: false,
+		},
+		{
+			Serial:    1004,
+			FQDNHash:  []byte("hash of C"),
+			DNS:       "c.example.org",
+			NotBefore: ctx.fc.Now().Add((-100 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((-10 * 24) * time.Hour),
+			IsRenewed: true,
+		},
+		{
+			Serial:    1005,
+			FQDNHash:  []byte("hash of C"),
+			DNS:       "c.example.org",
+			NotBefore: ctx.fc.Now().Add((-80 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((10 * 24) * time.Hour),
+			IsRenewed: true,
+		},
+		{
+			Serial:    1006,
+			FQDNHash:  []byte("hash of C"),
+			DNS:       "c.example.org",
+			NotBefore: ctx.fc.Now().Add((-75 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((15 * 24) * time.Hour),
+			IsRenewed: true,
+		},
+		{
+			Serial:    1007,
+			FQDNHash:  []byte("hash of C"),
+			DNS:       "c.example.org",
+			NotBefore: ctx.fc.Now().Add((-1 * 24) * time.Hour),
+			NotAfter:  ctx.fc.Now().Add((89 * 24) * time.Hour),
+			IsRenewed: false,
+		},
+	}
+
+	setupDBMap, err := sa.NewDbMap("mysql+tcp://test_setup@localhost:3306/boulder_sa_test")
+
+	for _, testData := range testCerts {
+		testData.stringSerial = strconv.Itoa(testData.Serial)
+
+		rawCert := x509.Certificate{
+			Subject: pkix.Name{
+				CommonName: testData.DNS,
+			},
+			NotBefore:    testData.NotBefore,
+			NotAfter:     testData.NotAfter,
+			DNSNames:     []string{testData.DNS},
+			SerialNumber: big.NewInt(int64(testData.Serial)),
+		}
+		certDer, _ := x509.CreateCertificate(rand.Reader, &rawCert, &rawCert, &testKey.PublicKey, &testKey)
```

Comment:
We should avoid ignoring errors, even in tests. I've run into a number of situations where a test was much harder to debug because of an ignored error.